polynomials/polynomial_finder.py

```diff
@@ -0,0 +1,181 @@
+        return string
+
+
+class PolynomialIdentifier:
+    """Polynomial Identifier class interpolates a potential polynomial between 2 SingleWord object"""
+
+    def __init__(self, first_word: SingleWord, second_word: SingleWord) -> None:
+        self.w_1 = first_word
+        self.w_2 = second_word
+        self.iteration_estimation = self.__iteration_estimator()
+
+    def __find_diff_timestamp(self) -> np.float64:
+        """Private function returning the difference of timestamp between
+        the 2 SingleWord objects w_1 and w_2"""
+
+        return abs(self.w_1.start_timestamp - self.w_2.start_timestamp)
+
+    def __iteration_estimator(self) -> int:
+        """Private function estimating the theorical LFSR number of iteration
+        between w_1 and w_2 SingleWord objects"""
+        diff_ts = self.__find_diff_timestamp()
+        return int(diff_ts / data_period)
+
+    def __init_LFSRs(self) -> list[LFSR]:
+        """Private function instantiating all 32 LFSRs with the 32 polynomials of
+        the LH 2.0 all starting at the word having the earliest timestamp"""
+
+        initial_LFSR_value = (
+            self.w_1.data
+            if self.w_1.start_timestamp < self.w_2.start_timestamp
+            else self.w_2.data
+        )
+        return [LFSR(poly, initial_LFSR_value) for poly in polys]
+
+    def search_polynomial(self) -> list[BeamAnalyzed]:
+        """Tries to interpolate a list of theorical polynomials for this w_1 and w_2 SingleWord object
+        Whole description of this algorithm on the Trello"""
+
+        found_polys = []
+        lfsrs = self.__init_LFSRs()
+        final_LFSR_value = (
+            self.w_1.data
+            if self.w_1.start_timestamp > self.w_2.start_timestamp
+            else self.w_2.data
+        )
+        for lfsr in lfsrs:
+            for _ in range(self.iteration_estimation - lfsr_iteration_approx):
+                lfsr.next()
+            for _ in range(2 * lfsr_iteration_approx):
+                if lfsr.next() == final_LFSR_value:
+                    found_polys.append(
+                        BeamAnalyzed(
+                            self.iteration_estimation,
+                            lfsr=lfsr,
+                            iteration_2_words=lfsr.cpt_for(final_LFSR_value),
+                        )
+                    )
+        if len(found_polys) == 0:
+            found_polys.append(BeamAnalyzed(self.iteration_estimation))
+
+        return found_polys
+
+
+bmc_decoder = BMC_decoder("../data/12MHz_2000ms")
+bmc_decoder.decode_whole_document()
+
+synthesis = SynthetizedBeamsAnalyzed()
+
+for i in range(len(bmc_decoder.indexes_0_envelope)):
+    words = bmc_decoder.get_first_and_last_word_from_beam(i)
+    if words != None:
+        polynomial_identifier1 = PolynomialIdentifier(words[0], words[1])
+        synthesis.add_beam(polynomial_identifier1.search_polynomial())
+
+synthesis.save_synthesis("synthesis.txt")
```

Comment:
This part shouldn't be in a package.

Reply:
[Done.](https://github.com/robotique-ecam/ViveTracker/pull/3/commits/dd334afd2b9466e6406a73d7271c79d1f1e5171a)